File: C++/Palindromes/break_Palindrome.cpp
// Break Palindrome
// Author: Pavan Kumar Paluri
// Time Complexity: O(N) and Space Complexity: O(1)
#include <iostream>
#include <string>
using namespace std;

class Solution{
public:
	string break_palindrome(string pal)
	{
		int len_pal = pal.size();
		for(int i=0; i<len_pal/2; i++)
		{
			// if a non-a: replace it
			if(pal[i]!='a')
			{
				pal[i]='a';
				return pal;
			}
			else
				continue;
		}

		// if all a's encountered, replace the last a with a b to obtain lexicographically smallest string
		if(pal.size()>1)
			pal[len_pal-1]='b';
		else 
			pal="";
		return pal;
	}
};

int main()
{
	Solution sol;
	cout << sol.break_palindrome("malayalam") << endl;
	return 0;
}

File: Python/Palindromes/Break_Palindrome.py
# Break A Palindrome
# Auhtor: Pavan Kumar Paluri
# LeetCode Question: https://leetcode.com/problems/break-a-palindrome/

class Solution:
    def breakPalindrome(self, palindrome: str) -> str:
        # break a pal
        for i in range(0, len(palindrome)//2):
            # If a non 'a' encountered in any pos of first half of pal, replace it with 'a'
            if palindrome[i] != 'a':
                return palindrome[:i]+'a'+palindrome[i+1:]
            else:
                continue
        # if all were 'a's in first half of string: replace last 'a' with 'b' to get lexicographically smallest non-pal string
        return palindrome[:-1]+'b' if len(palindrome)>1 else ""
